contracts/StorageManager.sol

```diff
@@ -14,87 +15,81 @@ contract StorageManager {
-     - availableCapacity: the amount of bytes currently offered. When the capacity is zero, already started Agreement can't be prolonged or re-started
-     - maximumDuration: the maximum time (in seconds) for which a customer can prepay.
+     - availableCapacity: the amount of bytes currently offered. When the capacity is zero, already started Agreement can't be prolonged or re-started ==> it is terminated
      - billingPlans: maps a billing period to a billing price. When a price is 0, the period is not offered.
      - agreementRegistry: the proposed and accepted Agreement
     */
     struct Offer {
         uint128 availableCapacity;
-        uint128 maximumDuration;
         mapping(uint64 => uint64) billingPlans;
         mapping(bytes32 => Agreement) agreementRegistry; // link to agreement that are accepted under this offer
     }
 
     /*
     Agreement represents:
-     - billingPrice: price per byte that is collected per each period. The contract can be cancelled by the proposer every duration seconds since the start.
-     - billingPeriod: period how often are funds collected.
+     - billingPrice: price per byte that is collected per each period.
+     - billingPeriod: period how often billing happens.
      - size: allocated size for the Agreement (in bytes, rounded up)
-     - startDate: when the Agreement was accepted
-     - numberOfPeriodsDeposited: number of periods (chosenPrice.duration seconds) that is deposited in the contracts.
-       At startDate * numberOfPeriodsDeposited seconds the Agreement expires unless topped up in the meantime
-     - numberOfPeriodsWithdrawn how many periods are withdrawn from the numberOfPeriodsDeposited. Provider can withdraw every period seconds since the start
-     - startDate: when Agreement was accepted
+     - availableFunds: funds available for the billing of the Agreement.
+     - lastPayoutDate: When was the last time Provider was payed out. Zero either means non-existing or terminated Agreement.
     */
     struct Agreement {
         uint64 billingPrice;
         uint64 billingPeriod;
-        uint64 numberOfPeriodsDeposited;
-        uint64 numberOfPeriodsWithdrawn;
+        uint256 availableFunds;
         uint128 size;
-        uint128 startDate;
+        uint128 lastPayoutDate;
     }
 
     // offerRegistry stores the open or closed Offer for provider.
     mapping(address => Offer) public offerRegistry;
 
     event AvailableCapacitySet(address indexed provider, uint128 capacity);
-    event MaximumDurationSet(address indexed provider, uint128 maximumDuration);
     event BillingPlanSet(address indexed provider, uint64 period, uint64 price);
     event MessageEmitted(address indexed provider, bytes32[] message);
 
     event NewAgreement(
+        bytes32 agreementReference,
         bytes32[] dataReference,
-        address indexed agreementer,
+        address indexed agreementAuthor, // TODO: [Q] Do we need information about who created the Agreement?
```

Comment:
[Q] Do we need information about who created the Agreement?

Reply:
Heh, damn. I meant in this event, but you are actually right that there is no check in the `withdrawFund` for who is calling it. Ups 😅

Reply:
Oh false alert 😅 Rinke thought about that and the `agreementReference` is calculated using the caller's address hence only he can modify the Agreement. Good thinking! Added a test for this just in case.

Sooo the question still stand. I don't think this is needed because you can just have a look on who called the `newAgreement` and that is `agreementAuthor` no need to specify it here?

Reply:
I think its good to have the "consumer" address in the Event. If somebody needs to consume this info from the event is a crucial piece of info (much better than trying to decrypt it from some other attribute). I would keep the field in the Event.